src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -903,6 +926,31 @@ def _create_schedule(name):
         '''
         return KernelSchedule(name)
 
+    def generate_psyir(self, parse_tree):
+        '''Translate the supplied fparser2 parse_tree into PSyIR.
+
+        :param parse_tree: the supplied fparser2 parse tree
+        :type parse_tree: :py:class:`fparser.two.Fortran2003.Program`
+
+        :returns: PSyIR representation of the supplied fparser2 parse_tree.
+        :rtype: :py:class:`psyclone.psyir.nodes.Container` or \
+            :py:class:`psyclone.psyir.nodes.Routine`
+
+        :raises GenerationError: if the root of the supplied fparser2 \
+            parse tree is not a Program
+
+        '''
+        if not isinstance(parse_tree, Fortran2003.Program):
+            raise GenerationError(
+                "The Fparser2Reader generate_psyir method expects root the "
+                "supplied fparser2 tree to be a Program, but found '{0}'"
+                "".format(type(parse_tree).__name__))
+        node = Container("dummy")
```

Comment:
Since we have guaranteed that the node is a Forrtan2003.Program can we just simply do:
```retrun self._program_handler(parse_tree, None)```

if this is be problematic, the dummy parent node is something we already use a few times in this front-end, so it is fine as a solution.

Reply:
I originally did `self.process_nodes(None, [parse_tree])` which does the same thing as you suggest. However, in either case the problem is that `process_nodes` always assumes that a PSyIR parent exists and tries to add children to it. Rather than changing the logic, I went with a dummy container. I think something like the following would work though if you would prefer that ...

```
                ...
                if psy_child:
                    if not parent:
                        if not len(nodes) == 1:
                            raise InternalError("xxx")
                        parent = nodes[0]
                    else:
                        self.nodes_to_code_block(parent, code_block_nodes)
                        parent.addchild(psy_child)
                ...
```

Reply:
I don't understand this `parent = nodes[0]`, parent is a PSyIR but nodes is fparser2 ast. Also process_nodes currently doesn't return anything, it just gives back the new children by appending them as the parent's children (which is why it always need some parent).

I was recommending the handler because handlers should always return a node (or give an exception). But now I see `_program_hanlder` doesn't. I believe program_handler should have  `return self._create_child(node.items[0], parent)` instead of the process_nodes call (similar to `_parenthesis_handler`). Would it work then?

Reply:
I know what I am proposing won't work for top-level CodeBlocks (?). So if it requires changing the code too much remember that I am OK with the dummy parent as well.

Reply:
After discussion, we have agreed to keep the use of a dummy node in this PR and to defer the potential restructuring of `process_nodes` as it would be a separate issue.